usb/ehci/ehci.c

```diff
@@ -66,24 +66,28 @@ static void ehci_stopAsync(hcd_t *hcd)
 
 static void ehci_qtdLink(ehci_qtd_t *prev, ehci_qtd_t *next)
 {
-	prev->hw->next = QTD_PTR(next);
+	prev->hw->next = next->paddr;
 	ehci_memDmb();
 }
 
 
-static void ehci_enqueue(ehci_qh_t *qh, ehci_qtd_t *first, ehci_qtd_t *last)
+static void ehci_enqueue(hcd_t *hcd, ehci_qh_t *qh, ehci_qtd_t *first, ehci_qtd_t *last)
 {
+	ehci_t *ehci = (ehci_t *)hcd->priv;
```

Comment:
This cast shouldn't really be necessary - you can use forward declaration to be able to hide this struct implementation in public header, if that's the reason

Reply:
The reason for it is not hiding implementation details, but to allow a kind of "polymorphism", so that there are different host controllers, which keep their context in this pointer. At the moment we only have ehci.